### ISSUE TITLE
Implement

### ISSUE THREAD

**Issue by sbinet** [gonum/gonum#328]
We are missing the numpy equivalent of `numpy.interp`:

Perhaps something along the lines of:

`Interpolate(x, xdata, ydata []float64) (y []float64)`

**Comment by kortschak** [comment]
Where? This looks like it probably fits in floats.

If it is implemented, it should also take a `dst []float64` argument reflected into `y`.

**Comment by btracey** [comment]
That's a good point, `floats` is a good fit.

I think instead should be 
```
Interpolate(x float64, xData, yData []float64) float64
```

The only good reason for `x` to be a slice is if the `x` data is sorted.  On the other hand, I think there are a lot of cases where one would only want the prediction at one `x`.

**Comment by btracey** [comment]
Also, the single input matches, for instance, `Nearest` and `NearestWithinSpan`

**Comment by kortschak** [comment]
Yeah, I was thinking single value x would work, just wasn't sure what the preamble would cost and whether that would be worth amortising over a set of x's. As a single value, it composes well with [Span](https://godoc.org/gonum.org/v1/gonum/floats#Span).

**Comment by btracey** [comment]
It depends if we're going to ensure xData and yData are sorted, or only panic if we find they are not. In big O notation they're the same either way, but different constants

**Comment by btracey** [comment]
We could have `Interpolate` and `InterpolateList` if we need it. In `InterpolateList` though `x` needs to be sorted, otherwise there's no point.

**Comment by kortschak** [comment]
I think requiring the input data to be sorted it pretty reasonable.

**Comment by sbinet** [comment]
It's borderline bikeshedding but what about a new package 'interp' ?
We could also put bezier and splines in there.

Somehow, it feels wrong to me to put that inside 'floats'.
I feel like 'floats' is becoming the catch-all package.

**Comment by tweeter0830** [comment]
How about something like the following? (I need to document and more thoroughly test)

```go
package interp

import (
	"fmt"
	"sort"
)

func panicIfNotSorted(data []float64, variableName string) {
	if !sort.Float64sAreSorted(data) {
		panic("Expected " + variableName +
			" to be sorted, but it is not.")
	}
}

func panicIfUnequalLength(x, y []float64, xName, yName string) {
	if len(x) != len(y) {
		panic(fmt.Sprintf(
			"Expected %s and %s to have the same length, "+
				"but found lengths of %d and %d instead.",
			xName, yName, len(x), len(y)))
	}
}

func interp1DBetweenPoints(x, x0, y0, x1, y1 float64) float64 {
	return y0 + (x-x0)*(y1-y0)/(x1-x0)
}

func linear1D(x float64, xData, yData []float64) (index int, y float64) {
	index = sort.SearchFloat64s(xData, x)
	switch index {
	case 0:
		y = yData[0]
	case len(xData):
		y = yData[len(yData)-1]
	default:
		y = interp1DBetweenPoints(
			x, xData[index-1], xData[index],
			yData[index-1], yData[index])
	}
	return
}

func Linear1D(x float64, xData, yData []float64, checkSorted bool) (y float64) {
	panicIfUnequalLength(xData, yData, "xData", "yData")
	if checkSorted {
		panicIfNotSorted(xData, "xData")
	}
	_, y = linear1D(x, xData, yData)
	return
}

func LinearList1D(x, xData, yData []float64, checkSorted bool) (y []float64) {
	panicIfUnequalLength(xData, yData, "xData", "yData")
	if checkSorted {
		panicIfNotSorted(x, "x")
		panicIfNotSorted(xData, "xData")
	}
	y = make([]float64, len(x))
	lastIndex := 0
	for idx, xVal := range x {
		lastIndex, y[idx] = linear1D(
			xVal, xData[lastIndex:], yData[lastIndex:])
	}
	return
}
```

I like the idea of an interp package that eventually incorporates multidimensional interpolations as well as spline interpolation.

**Comment by jonreiter** [comment]
i've had a go at this in my interp branch [here](https://github.com/jonreiter/gonum/tree/interp).

it's built around an interp.Algo interface for interpolators with Value() and Values() methods.  for linear sorting is enough but for splines -- or anything else non-trivial -- its more reasonable to have a fit method and to store state.  i've done linear, step and cubic spline -- which i needed anyway -- to show how that works.

thoughts on this approach?  it doesn't look exactly like anything discussed above but i think is more composable this way.  removing fit feels better in some ways but that requires copying over the input data.
